jsapp/js/bem.ts

```diff
@@ -1,382 +1,103 @@
+interface BemComponentProps extends React.ComponentProps<any> {
+  m?: string | string[] | BemModifiersObject
+}
 
-bem.BackgroundAudioPlayer = bem.create('background-audio-player');
-bem.BackgroundAudioPlayer__label = bem.BackgroundAudioPlayer.__('label', '<label>');
-bem.BackgroundAudioPlayer__audio = bem.BackgroundAudioPlayer.__('audio', '<audio>');
+interface BemInstance extends React.ComponentClass<BemComponentProps, {}> {
+  blockName: string
+}
 
-export default bem;
+/**
+ * Creates a BEM class for block or element.
+ * For first parameter pass `null` to create a Block component,
+ * or pass existing Block component to create Element component (a child).
+ * Please no angle brackets for `htmlTagName`.
+ */
+export function makeBem(
+  parent: BemInstance | null,
+  name: string,
+  htmlTagName: string = 'div'
+): BemInstance {
+  class BemComponent extends React.Component<BemComponentProps, {}> {
+    static blockName: string = parent ? parent.blockName : name
+
+    constructor(props: BemComponentProps) {
+      super(props);
+    }
+
+    render() {
+      let classNames: string[] = [];
+
+      // Keep existing className attribute if given (either string or array)
+      if (typeof this.props.className === 'string') {
+        classNames.push(this.props.className);
+      } else if (Array.isArray(this.props.className)) {
+        classNames.push(this.props.className.join(' '));
+      }
+
+      // wholeName includes parent, e.g. `parent-block__child-element`
+      const wholeName = parent ? `${parent.blockName}__${name}` : name
+      classNames.push(wholeName);
+
+      // modifiers could be a single string or array of strings or object with booleans
+      if (typeof this.props.m === 'string' && this.props.m.length >= 1) {
+        // Case 1: string
+        classNames.push(`${wholeName}--${this.props.m}`);
+      } else if (Array.isArray(this.props.m)) {
+        // Case 2: array
+        this.props.m.forEach((modifier) => {
+          if (typeof modifier === 'string' && modifier.length >= 1) {
+            classNames.push(`${wholeName}--${modifier}`);
+          }
+        });
+      } else if (typeof this.props.m === 'object') {
+        // Case 3: object
+        Object.entries(this.props.m).forEach((entry) => {
+          if (entry[1] === true) {
+            classNames.push(`${wholeName}--${entry[0]}`);
+          }
+        });
+      }
+
+      const newProps: {[propName: string]: any} = {
+        className: classNames.join(' ')
+      }
+
+      // Keep all the original props expect for modifiers (don't need it) and
+      // className (we use our own).
+      Object.entries(this.props).forEach((propEntry) => {
+        if (propEntry[0] !== 'm' && propEntry[0] !== 'className') {
+          newProps[propEntry[0]] = propEntry[1];
+        }
+      });
+
+      return React.createElement(htmlTagName, newProps);
+    }
+  }
+
+  return BemComponent
```

Comment:
Is the class definition being returned here? Or an instance of the class? To me it looks like this function reads:
```
function(...) {
  class Foo {...};
  return Foo;
}
```
and I'm a little confused

Reply:
Besides these, this method of bem was :sparkles: clean to read through :+1:

Reply:
This returns a class, not an instance. Or speaking differently it returns a component, i.e. the `bem.TextBox` is a React component. So when calling `makeBem()` you are making a new class/component to be used later. And when writing JSX code `<bem.TextBox>…</bem.TextBox>` React takes the provided class and renders it.

Reply:
So it's:

```
function(someParams) {
  class Foo {/* configure it using `someParams` making the returned `Foo` unique */};
  return Foo;
}
```

Reply:
Gotcha! well explained :)